### ISSUE TITLE
v1.1.1 introduce breaking change for stub query strings

### ISSUE THREAD

**Issue by baelter** [excon/excon#870]
Writing up a test now. Will update issue asap

**Comment by geemus** [comment]
Ok, just let me know. I thought the changes should only impact mock usage, but we certainly may have missing test coverage there still.

**Comment by baelter** [comment]
Indeed it should only affect stubs. But still breaking :) 
`+` are no longer preserved.

```ruby
require "bundler/inline"

gemfile do
  source "https://rubygems.org"

  gem "excon", "1.1.1"
end

puts Excon::VERSION
puts Excon.stub({ query: "a+b" }, {})[0]
```

>

```
1.1.0
{:query=>{"a b"}}
```

```ruby
require "bundler/inline"

gemfile do
  source "https://rubygems.org"

  gem "excon", "1.1.0"
end

puts Excon::VERSION
puts Excon.stub({ query: "a+b" }, {})[0]
```

>

```
1.1.1
{:query=>{"a+b"}}
```

**Comment by baelter** [comment]
Hm, `+` is preserved in `0.112.0` so this change in behaviour is only for `1.0.0` and `1.1.0`, Introduced by https://github.com/excon/excon/commit/d8365526c369ef28dcee81bca20bc72074eacccc

**Comment by geemus** [comment]
@baelter Yep, definitely still breaking. I just wanted to point out that it was mocks specifically in case that helped narrow down the test case. I'll take a look presently.

**Comment by geemus** [comment]
Are you relying directly on the return value from `Excon.stub` somehow? Or is this failing to match when you call it? Just trying to make sure I understand. I thought the query value that you show there was really an internal value, so I knew it would change but I didn't think it would impact end users. Want to make sure I'm chasing the right bug/fix.

**Comment by baelter** [comment]
This also breaks matching: https://github.com/excon/excon/compare/v0.112.0...master#diff-870d481afe6dde6869e2e7aaa8586034283a992a04f4ed683ade71b99ef7ddf3R157
since it adds `'Accept-Encoding'=>'deflate, gzip'` to all requests.

**Comment by geemus** [comment]
Thanks for the clarification, the webmock part does complicate things a bit (and isn't something I'm as familiar with).

**Comment by baelter** [comment]
Maybe this can help: https://github.com/baelter/excon-webmock-issues

**Comment by baelter** [comment]
The other issue I saw was not related to the `Decompress` middleware but the default port change: https://github.com/excon/excon/commit/a4b5b44c8428d79a0a507efe7b05df65192646b2

**Comment by dentarg** [comment]
The default port change is mentioned in the changelog: https://github.com/excon/excon/blob/161d414d55649e7f392c93f9b287f857d8963e34/changelog.txt#L16

Finally the 10 year old issue could be closed :)  https://github.com/excon/excon/issues/357

**Comment by geemus** [comment]
@baelter - thanks, those are quite helpful.

I think this addresses the escaping issue: https://github.com/excon/excon/pull/871

I'll work on the headers next.

Could you provide more context on what is going wrong for you with default ports? I don't think the cases you shared run into that.

**Comment by geemus** [comment]
Ok, after looking at headers some more, I think it's probably all actually working as intended (it is a breaking change, but happened over a major version boundary, fwiw).

Basically, in 1.0.0 we added the decompress middleware. It automatically adds headers to ask for compressed data during the request and then does the decompression of the response if needed. It looks like there were previously other middlewares that also manipulated headers, fwiw, so this behavior isn't totally new/unheard of (though it is new as the default).

I think the correct fix is probably to either disable the decompress middleware in your usage, if you don't desire it and want to return to the pre-1.0 behavior. Or to add that header to the mocks so that they reflect what is actually sent over the wire. Does that makes sense?

Alternatively, I considered whether the current matching is overly aggressive. In particular, I could imagine switching from matches only triggering if the request has all the headers the stub has to something where the request matches if all of it's supplied headers match the ones in the stubs. Unfortunately, this wouldn't help here anyway because webmock hijacks things and uses it's own lookup/handler:

https://github.com/bblimke/webmock/blob/master/lib/webmock/http_lib_adapters/excon_adapter.rb#L41-L43

Still trying to wrap my head around exactly what how webmock is integrating too (and how that relates or not to excon's own mocking)...

**Comment by geemus** [comment]
Here is (I think) the equivalent to the repro cases, just using excon's mocks directly, which passes:

```
require 'bundler/inline'

require './lib/excon'

gemfile do
  source 'https://rubygems.org'

  gem 'minitest'
end

require 'excon'
require 'minitest/autorun'

class TestExcon < Minitest::Test
  def test_header_internal
    Excon.stub(
      {
        headers: { 'accept' => 'application/json', 'host' => 'example.com:443' },
        method: :get,
        url: 'http://example.com'
      },
      { status: 200 }
    )
    http = Excon.new(
      'https://example.com',
      headers: { 'accept' => 'application/json' }
    )
    assert_equal 200, http.get.status
  end
end
```

So maybe no change is needed there. I guess it's maybe a broader question of webmock settings and/or strictness in matching?

**Comment by geemus** [comment]
Bah, never mind. That omitted `mock: true` in the params so it was making a real request to example.com :(

**Comment by geemus** [comment]
This example does appear to work:
```
require 'bundler/inline'

require './lib/excon'

gemfile do
  source 'https://rubygems.org'

  gem 'minitest'
end

require 'excon'
require 'minitest/autorun'

class TestExcon < Minitest::Test
  def test_header_internal
    Excon.stub(
      {
        headers: { 'accept' => 'application/json' },
        method: :get,
        url: 'http://example.com'
      },
      { status: 200 }
    )
    http = Excon.new(
      'http://example.com',
      headers: { 'accept' => 'application/json' },
      mock: true
    )
    assert_equal 200, http.get.status
  end
end
```

I'll note that I omitted the explicit host header for the stub. It's implied by the url and supplying both introduces some ambiguity about which should be correct. At least in excon's internal usage, the mocks still match even without it. So it seems webmock is more strict about it's expectations for matching (and that seems to be out of my control). Does that make sense?

**Comment by baelter** [comment]
>Still trying to wrap my head around exactly what how webmock is integrating too (and how that relates or not to excon's own mocking)...

Me to :)
`Excon.stubs.clear` seems to interfere with webmock though.

Thanks for digging into this. I'll see if someone from webmock can shine some light on this.

**Comment by geemus** [comment]
@baelter Thanks. Would be great to hear their insights since I can make some educated guesses, but it's a bit hard beyond that. I think that webmock is relying on one particular stub (which matches all requests and hands off to webmock for processing), so clearing would likely break that.

**Comment by geemus** [comment]
Hoping the query portion is now fixed in v1.2.0. But please let me know if that's not the case. There may still be work to be done on default ports though, also please let me know.

**Comment by baelter** [comment]
Confirmed! port is not included by default, which is fine. I just updated our mocks. Thanks a lot!
